fss_check/fss_utils.py

```diff
@@ -0,0 +1,616 @@
+    out = dok["date", "pitch", "roll", "kalman"]
+    out["roll_err"] = dok["roll_fss"] - dok["roll"]
+    out["pitch"].format = ".2f"
+    out["roll"].format = ".2f"
+    out["roll_err"].format = ".2f"
+    return out
+
+
+def get_fss_data_intervals(
+    dat,
+    mask=None,
+    sun_presence=None,
+    pitch_in_fov=None,
+    pitch=None,
+    dpitch=1.0,
+    roll=None,
+    droll=1.0,
+    pitch_err=None,
+    dpitch_err=1.0,
+    roll_err=None,
+    droll_err=1.0,
+    dt_join=1000,
+):
+    """Get intervals of FSS data within boxes of pitch or roll
+
+    Parameters
+    ----------
+    dat : numpy.ndarray or astropy.table.Table
+        Table of FSS data
+    mask : numpy.ndarray or None
+        Boolean mask to apply to ``dat``.
+    sun_presence : bool, None
+        If not None select with matching sun presence.
+    pitch_in_fov : bool, None
+        If not None select data with pitch in the FSS FOV.
+    pitch : float or None
+        Select data within ``dpitch`` of ``pitch`` value if supplied (deg).
+    dpitch : float
+        Pitch width threshold (deg)
+    roll : float or None
+        Select data within ``droll`` of ``roll`` value if supplied (deg).
+    droll : float
+        Roll width threshold (deg)
+    pitch_err : float or None
+        Select data within ``dpitch_err`` of ``pitch_err`` value if supplied (deg).
+    dpitch_err : float
+        Pitch error outlier threshold (deg)
+    roll_err : float or None
+        Select data within ``droll_err`` of ``roll_err`` value if supplied (deg).
+    droll_err : float
+        Roll error outlier threshold (deg)
+    dt_join : float
+        Join intervals separated by less than ``dt_join`` (sec)
+
+    Returns
+    -------
+    intervals : Table of (start, stop) tuples
+        Intervals of FSS data that are outliers in pitch or roll
+    """
+    from cheta.utils import logical_intervals
+    from kadi.events.models import fuzz_states
+
+    ok = np.ones(len(dat), dtype=bool)
+    if mask is not None:
+        ok &= mask
+    if sun_presence is not None:
+        sp = dat["alpha_sun"] & dat["beta_sun"]
+        ok &= sp if sun_presence else ~sp
+    if pitch_in_fov is not None:
+        pif = dat["pitch"] < 135
+        ok &= pif if pitch_in_fov else ~pif
+    if pitch is not None:
+        ok &= np.abs(dat["pitch"] - pitch) < dpitch
+    if roll is not None:
+        ok &= np.abs(dat["roll"] - roll) < droll
+    if pitch_err is not None:
+        ok &= np.abs(dat["pitch_fss"] - dat["pitch"] - pitch_err) < dpitch_err
+    if roll_err is not None:
+        ok &= np.abs(dat["roll_fss"] - dat["roll"] - roll_err) < droll_err
+
+    intervals = logical_intervals(dat["times"], ok, max_gap=33)
+    intervals = fuzz_states(intervals, dt_join)
+    intervals["duration"].format = ".1f"
+    intervals["tstart"].format = ".1f"
+    intervals["tstop"].format = ".1f"
+
+    return intervals
+
+
+def get_spm_pitch_roll():
+    """Get pitch/roll values for Sun Position Monitor.
```

Comment:
I don't remember, is this something that will change with updates?

Reply:
It would change if there were an OBC flight software patch to the sun position monitor
region boundaries. This is pretty unlikely but good to keep in mind.